plugins/module_utils/validation.py

```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021, XLAB Steampunk <steampunk@xlab.si>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+def missing_from_params_and_remote(params, module_params, record=None):
+    missing = []
+    for p in params:
+        if module_params[p] is not None or (record and record[p]):
```

Comment:
This is safe to perform if we assume that all parameters are strings. Under that assumption, `module_params[p]` and `record[p]` can only be `None`, `""`, or some other string, and things will work properly. But if we ever get to add a boolean parameter, things could start triggering when we do not want them to. So, maybe we should document our assumptions at the start of the function (and maybe even add a check that all values that we are using are `None`s or strings)?

Another assumption this function makes is that the record, if present, will always contain values for all parameters we are checking. Again, this is true as far as we know because the REST API always returns all record fields (the unset ones are set to the default value of `""`), but maybe we should document this fact also? Or even add a check like `if not set(params).issubset(record): raise ...` that will catch changes in the ServiceNow's API? Assuming that all parameters are in `module_params` is OK because inverse situation would mean we have a bug in our code.

Reply:
Thanks for pointing this out. I equipped the code with additional checks and documentation, so that it's clearer what the assumptions are. I'm still not particularly fond of this piece of code, but I guess it'll do for now.